### PR TITLE
Some (sadly breaking) improvements to the Fluid API

### PR DESCRIPTION
**Pull request by gigaherz**
* Replaced the string id on the FluidAttributes with a reference to the owner fluid.
* Added a boolean to distinguish when fluids want to use biome-based water coloring.
* Added logic to automatically gather fluid textures into the list of textures to bake.

### CODE REVIEW

**Comment by pupnewfster** [comment]
I am not positive, but I think water and lava *should* override the translation key as I don't believe there are any lang keys in vanilla for fluids that start with `fluid`. I do agree that the way this PR is introducing for the default translation key is better though.

Edit: I am blind already is overridden.

**Comment by Ommina** [comment]
Is the expectation here that the mod author will create a new class extending FlowingFluid for each and every fluid they create?  It feels like any mod that adds more than a couple fluids are going to rapidly get lost in a sea of copy/paste code.

I've been working on generalize FlowingFluid, but it isn't pretty, and the change to FluidAttributes taking a Fluid reference in this PR is going to make it even less so.

**Comment by gigaherz** [comment]
Ultimately, yes, like we do for blocks and items. I'm working toward making it possible to create a more flexible system using suppliers, but my focus is on the base work in building up the vanilla system.

I was considering making something equivalent to Item.Properties and Block.Properties, for constructing a FlowingFluid (would have to be a new ForgeFlowingFluid class) so that implementing the new fluids doesn't require spelling out all the boilerplate, but that will come later.

File: patches/minecraft/net/minecraft/block/FlowingFluidBlock.java.patch
Comment (Yamahari) [inline patches/minecraft/net/minecraft/block/FlowingFluidBlock.java.patch]: while you're at it, this could be simplified to 
`p_200122_2_.func_204520_s().func_206886_c().func_207187_a(getFluid()) || super.func_200124_e(p_200122_1_);`

Comment (gigaherz) [reply patches/minecraft/net/minecraft/block/FlowingFluidBlock.java.patch]: It could, but I think that would also make the patch less obvious, I'd rather leave it as it is, so that people see what REALLY changed in that patch.

File: src/main/java/net/minecraftforge/fluids/FluidAttributes.java
Comment (mezz) [inline src/main/java/net/minecraftforge/fluids/FluidAttributes.java:164]: to avoid a breaking change here, can this become
```java
return owner.getRegistryName().toString();
```
and deprecate it?

File: src/main/java/net/minecraftforge/fluids/FluidAttributes.java
Comment (mezz) [inline src/main/java/net/minecraftforge/fluids/FluidAttributes.java]: This flag is too specific to be here, it is not at the right level of abstraction.

Instead, I think we should create subclass of `FluidAttributes` just for water that overrides `FluidAttributes#getColor(IEnviromentBlockReader, BlockPos)` to call `BiomeColors.getWaterColor`.
That should simplify the patch a little as well.

**Review by mezz — CHANGES_REQUESTED** [review]

File: src/main/java/net/minecraftforge/fluids/FluidAttributes.java
Comment (KingLemming) [reply src/main/java/net/minecraftforge/fluids/FluidAttributes.java:164]: Do we actually have an RB at this point? I think we're early enough in dev that we shouldn't have to overly worry about deprecation. I'd rather it break and people use it correctly.

Comment (gigaherz) [reply src/main/java/net/minecraftforge/fluids/FluidAttributes.java:164]: That is my thinking. On top of that, this PR would have breaking changes regardless, so the value of this deprecated method that I just added last week seems too low to be worth keeping it.

I was hoping to get this merged quickly to avoid a lot of people having to deal with re-releasing updates, but I really want Lex's input on this.

File: src/main/java/net/minecraftforge/fluids/FluidAttributes.java
Comment (gigaherz) [reply src/main/java/net/minecraftforge/fluids/FluidAttributes.java]: I was initially thinking I don't like the idea, but actually, after poking at it in the code, I do.

It actually doesn't change the patch that much, but I guess the logic in FluidBlockRenderer is simpler now.

File: src/main/java/net/minecraftforge/fluids/FluidAttributes.java
Comment (Alpvax) [inline src/main/java/net/minecraftforge/fluids/FluidAttributes.java]: Could we not add a simpler way of subclassing the attributes without subclassing the builder. e.g. by optionally passing in a factory to `build` (i.e. the water builder would become `build(Water::new)`) or to the builder constructor, or adding a new builder function `setAttributesClass` or something.
That would make it easier for other mods which wanted to add a custom colour map, or otherwise override one of the FluidAttributes implementations.

File: src/main/java/net/minecraftforge/fluids/FluidAttributes.java
Comment (KingLemming) [reply src/main/java/net/minecraftforge/fluids/FluidAttributes.java:164]: Yeah, if changes are *already* broken then I say maximize all possible breakage now. Let's make 1.14 as clean as possible. No deprecations, etc. Damn the torpedoes.

Comment (mezz) [reply src/main/java/net/minecraftforge/fluids/FluidAttributes.java:164]: Ok, makes sense to me

**Comment by ichttt** [comment]
is this done? Would be a good time to merge now, as we just had another breaking change

**Comment by gigaherz** [comment]
It is. We haven't had any other fluids issue. I just wanted Lex's confirmation that this is good to merge, since it adds quite a few patches to 2 classes.

**Comment by LexManos** [comment]
I was waiting on you to tell me things are ready to look at.

File: src/main/java/net/minecraftforge/fluids/FluidAttributes.java
Comment (LexManos) [reply src/main/java/net/minecraftforge/fluids/FluidAttributes.java:164]: We are holding off a RB explicitly for this fluid system, so that it can do everything correctly and THEN we can stabilize the binary version. So once this is done, we'll worry about compat.

File: patches/minecraft/net/minecraft/fluid/Fluid.java.patch
Comment (LexManos) [inline patches/minecraft/net/minecraft/fluid/Fluid.java.patch:28]: This would be better off moving to a ForgeHooks method, just due to the shear size, and uglyness that is due to being in a patch.

File: patches/minecraft/net/minecraft/block/FlowingFluidBlock.java.patch
Comment (LexManos) [inline patches/minecraft/net/minecraft/block/FlowingFluidBlock.java.patch]: Use our field to method trick we use for ItemStack/PotionEffect for this. Would vastly simplify the patch set and maintainability.

File: patches/minecraft/net/minecraft/item/BucketItem.java.patch
Comment (LexManos) [inline patches/minecraft/net/minecraft/item/BucketItem.java.patch]: Perhaps would be worth doing the same runtime redirect here as well to save a lot of patches

File: src/main/java/net/minecraftforge/fluids/FluidAttributes.java
Comment (LexManos) [inline src/main/java/net/minecraftforge/fluids/FluidAttributes.java]: javadocs are helpful on PUBLIC methods, so add this information to the getter as well.

File: src/main/java/net/minecraftforge/fluids/FluidAttributes.java
Comment (LexManos) [reply src/main/java/net/minecraftforge/fluids/FluidAttributes.java]: This should be a simple enough change to add, you should do it.

File: src/main/java/net/minecraftforge/fluids/FluidAttributes.java
Comment (LexManos) [inline src/main/java/net/minecraftforge/fluids/FluidAttributes.java]: Why does this need to be the fluid instance? I can see this being a chicken-egg style thing. 
Have you tried using a registry delegate or something that isn't the Fluid directly.
A delegate would silently support registry replacements. But doesn't solve the chicken/egg thing off hand.
All you seem to use it for is the name, what was the logic behind this change?

**Review by LexManos — CHANGES_REQUESTED** [review]

File: src/main/java/net/minecraftforge/fluids/FluidAttributes.java
Comment (gigaherz) [reply src/main/java/net/minecraftforge/fluids/FluidAttributes.java]: My thinking was that it may be useful for the people subclassing FluidAttributes, but on second thought, all methods that are worth overriding either have a FluidStack or a FluidState, meaning both have a .getFluid() method available, so the field could be simplified to a ResourceLocation.

**Comment by tterrag1098** [comment]
I would like to have a bit more discussion around the "fuel time" feature before we move forward. 

The main problem I see with this feature is that "fuel time" is not an intrinsic property of a fluid, unlike temperature or viscosity. How a given generator creates power from a liquid is totally open ended. So I propose retooling this feature as a more intrinsic property with a better defined meaning.

It could instead be defined as "energy density" i.e. the max amount of energy stored in the fluid, perhaps represented in FE/mB. This would define only the *maximum* amount of energy that can be generated by a fluid, but not *how* that energy is generated. For instance both lava and gasoline could implement this data point, but it's up to the generator to decide if they can use the specific fluid, perhaps by using tags.

**Comment by LexManos** [comment]
Then it's just a naming change. As it was intended that 'fuelTime' was only meant to be a seed value, it was up to other mechanics to determine if it could be burnt, or exploded, or chilled, or whatever mechanic you use to generate the energy.
The problem with extending things to FE/mB is that it then adds a constant value to furnace. And that's a odd conversion.
But as a community we already have the standard and use case of fluids as a furnace fuel. 
So I would still represent it as the 2000 lava has, the 'burntime' and let the consumers define their ratios for length/FE/whatever.

File: patches/minecraft/net/minecraft/fluid/Fluid.java.patch
Comment (gigaherz) [reply patches/minecraft/net/minecraft/fluid/Fluid.java.patch:28]: Done.

File: patches/minecraft/net/minecraft/block/FlowingFluidBlock.java.patch
Comment (gigaherz) [reply patches/minecraft/net/minecraft/block/FlowingFluidBlock.java.patch]: Done (tentative).

File: src/main/java/net/minecraftforge/fluids/FluidAttributes.java
Comment (gigaherz) [reply src/main/java/net/minecraftforge/fluids/FluidAttributes.java]: Removed the fuelTime stuff instead.

File: src/main/java/net/minecraftforge/fluids/FluidAttributes.java
Comment (gigaherz) [reply src/main/java/net/minecraftforge/fluids/FluidAttributes.java]: Removed the pointless field. The builder function still takes a Fluid, for convenience, but it's only used to get the registry name from.

**Comment by gigaherz** [comment]
At this point, this should be working, and could get a new pass of reviews.
The FlowingFluidBlock and BucketItem taking a supplier is done with the assumption that certain changes to `RegistryObject<T>` make it into the code, if not, those constructors may need to be reconsidered.

File: src/main/java/net/minecraftforge/common/ForgeHooks.java
Comment (pupnewfster) [inline src/main/java/net/minecraftforge/common/ForgeHooks.java]: Densty is set twice in this line

File: src/main/java/net/minecraftforge/fluids/FluidAttributes.java
Comment (pupnewfster) [inline src/main/java/net/minecraftforge/fluids/FluidAttributes.java]: As owner is only used for getting the registry name for the translation key, would it make sense to have two constructors, one that takes Fluid and one that takes a ResourceLocation (being the registry name). I don't fully remember the scope of how FluidAttributes gets made/used so this may not even make sense to do. (Or maybe it makes sense to change the protected builder to *only* ResourceLocation for registry name and make the `public static Builder builder(` above pass the registry name to this constructor.

File: src/main/java/net/minecraftforge/common/ForgeHooks.java
Comment (gigaherz) [reply src/main/java/net/minecraftforge/common/ForgeHooks.java]: oops!

File: src/main/java/net/minecraftforge/fluids/FluidAttributes.java
Comment (gigaherz) [reply src/main/java/net/minecraftforge/fluids/FluidAttributes.java]: `FluidAttributes` construction is something that should only happen within the `createAttributes` method of the Fluid class, so in principle 100% of the times the given value in that parameter will be the `this`, within `createAttributes`. The main reason the builder isn't given privately to the createAttributes method directly, is that mod fluids might want to construct `FluidAttributes` instances with extra parameters not represented in the default attributes.

**Comment by alexbegt** [comment]
This isn't that important but should the comment in the fieldtomethodtransformers.js say 'potion' for the Fluid block and Bucket item?

**Comment by gigaherz** [comment]
No, no it shouldn't. :P

File: src/main/resources/META-INF/fieldtomethodtransformers.js
Comment (ichttt) [inline src/main/resources/META-INF/fieldtomethodtransformers.js]: if getFluid is a method that has no srg name (i.e. added by a forge patch),  you can just call `asmapi.redirectFieldToMethod(classNode, fn, 'getFluid')`, as it does not need to be remapped

File: src/main/resources/META-INF/fieldtomethodtransformers.js
Comment (ichttt) [inline src/main/resources/META-INF/fieldtomethodtransformers.js]: same here

**Review by ichttt — CHANGES_REQUESTED** [review]
Sorry, some nitpicks from me regarding the asm as well

File: src/main/resources/META-INF/fieldtomethodtransformers.js
Comment (gigaherz) [reply src/main/resources/META-INF/fieldtomethodtransformers.js]: Oops, completely missed that lol.
/me hides in shame
Will fix when I'm back home later

**Comment by gigaherz** [comment]
I have added what I think is a really handy helper class for implementing fluids. I'm now going to sleep, but I plan on addressing any concern when I wake up, and then merge this PR, so if you have any comment abut the ForgeFlowingFluid, now's the time.

File: src/main/java/net/minecraftforge/fluids/ForgeFlowingFluid.java
Comment (pupnewfster) [inline src/main/java/net/minecraftforge/fluids/ForgeFlowingFluid.java]: Don't have my compiler open at the moment, so don't remember what this method is for overall given the naming, but is checking against water the correct thing to be doing here given its a custom fluid.

File: src/main/java/net/minecraftforge/fluids/ForgeFlowingFluid.java
Comment (pupnewfster) [inline src/main/java/net/minecraftforge/fluids/ForgeFlowingFluid.java:153]: Should this also compare any tags that this fluid is in and the given fluid is in? Or is that something that should be in an overwritten method.

File: src/main/java/net/minecraftforge/fluids/ForgeFlowingFluid.java
Comment (gigaherz) [reply src/main/java/net/minecraftforge/fluids/ForgeFlowingFluid.java]: TBH I meant to look at what this is used for, but I forgot, will check.

File: src/main/java/net/minecraftforge/fluids/ForgeFlowingFluid.java
Comment (gigaherz) [reply src/main/java/net/minecraftforge/fluids/ForgeFlowingFluid.java:153]: Although it may be interesting for a fluid to be able to say eg. that it's "just water" and it should be considered water by any neighbouring water blocks, such as for the level calculations, I don't think it would be correct to just compare tags, since there can be tags that mean something completely different, and we can't tell the difference in code.
I say if someone wants a different test than default, they can override isEquivalentTo.

File: src/main/java/net/minecraftforge/fluids/ForgeFlowingFluid.java
Comment (gigaherz) [reply src/main/java/net/minecraftforge/fluids/ForgeFlowingFluid.java]: Did my best. Settled on using isEquivalentTo as default. Mods can override if needed.

**Review by Alpvax — COMMENTED** [review]
JavaDoc for `FluidAttributes#getTranslationKey(FluidStack)` is incorrect, referring to "getUnlocalizedName" instead of "getTranslationKey".

Do we want to add some of the functions from (jabelar's) 1.12 PR to make fluids more/less water-like? i.e. drowning, watering plants, spawning aquatic creatures etc.
Or should we leave that for a different PR?

**Comment by gigaherz** [comment]
Extra features are out of the scope of this PR, they can be added on later, the RB is already late enough.
The goal here is to have a solid base that won't require breaking changes.

As for the javadoc, thanks, fixed.

**Pull request merged**
